app/src/main/java/com/github/h3lp3rs/h3lp/forum/implementation/CachedForum.kt

```diff
@@ -0,0 +1,172 @@
+package com.github.h3lp3rs.h3lp.forum.implementation
+
+import android.os.Build
+import androidx.annotation.RequiresApi
+import com.github.h3lp3rs.h3lp.forum.*
+import com.github.h3lp3rs.h3lp.forum.data.ForumPostData
+import com.github.h3lp3rs.h3lp.storage.Storages.*
+import com.github.h3lp3rs.h3lp.storage.Storages.Companion.storageOf
+import java.util.concurrent.CompletableFuture
+import java.util.stream.Collectors.toList
+
+/**
+ * This wrapper class forwards all calls to a Forum implementation and adds an additional
+ * layer of caching. This way, negative responses due to connectivity issues still resolve
+ * as valid old data output.
+ * NOTE: This cache behaves according to the ForumProtocol.md limitations (subset of FireForum
+ * interface)
+ * @param forum The wrapped forum we want to cache
+ */
+class CachedForum(private val forum: Forum) : Forum {
+    override val path = forum.path
+    private val cache = storageOf(FORUM_CACHE)
+
+    // Entry for all the posts at one path
+    data class CacheEntry(val posts: ArrayList<ForumPostData>)

```

Comment:
Mhhh this may work, but ArrayLists are safe values for serialization. 
Probably that would be smarter though if I can find an implemenation of a map that is easily serializable.

Reply:
I started changing it but I remember that the reason I did not do it was in reality the fact that all our forum structures takes lists, so I wanted to stay consistent (and don't bother with set to list (and vice versa) transformations.